blog/2024-04-29-secure-sveltekit.mdx

```diff
@@ -0,0 +1,123 @@
+  <OIDC />
+</details>
+
+## Adding a Non-Admin User
+
+<details>
+  <summary>Instructions</summary>
+  <NonAdminUser />
+</details>
+
+## Setting up a SvelteKit Project
+
+:::info
+We will use the Phase Two SvelteKit example code here, but the logic could easily be applied to any existing application.
+:::
+
+1. Clone the Phase Two [example repo](https://github.com/p2-inc/examples/).
+2. Open the SvelteKit [folder](https://github.com/p2-inc/examples/tree/main/frameworks/sveltekit) within `/frameworks/sveltekit`.
+3. Run `npm install` and then `npm run dev`. This example leverages [@auth/sveltekit](https://github.com/nextauthjs/next-auth) to provide HOC support.
+4. The project makes use of the following SvelteKit items: hooks, components, layout, and `@auth/sveltekit` module. We'll review each in kind.
+5. We'll review where we configure out Keycloak instance. Open the `src/auth.ts` file. This is a server only file. We will be updating a few values from the prior section where we set up our OIDC client. Taking the values from the OIDC Client Config section, set those values in the code. While it is recommended to use Environment variables for the secret, for the purpose of this tutorial, paste in the **Client secret** from the OIDC client creation section for the value of `clientSecret`. 
+
+```tsx
+// Use Environment Variables AUTH_SECRET in prod
+const authjsSecret = "f18d48ce9bea32e44b5591b2c89185729d4559435f77ca76872a83a0850563a4";
+
+const realm = "shared-deployment-001";
+
+const kcConfig = {
+  // Use Environment Variables AUTH_KEYCLOAK_ISSUER in prod
+  issuer: `https://usw2.auth.ac/auth/realms/${realm}`,
+  // Paste "Client id" here. Use Environment Variables AUTH_KEYCLOAK_ID in prod
+  clientId: "reg-example-1",
+  // Paste "Client secret" here. Use Environment Variables AUTH_KEYCLOAK_ISSUER in prod
+  clientSecret: "CLIENT_SECRET",
+};
+```
+
+Those are used to popluate the config for the provider `Keycloak`:
+
+```tsx
+export const { handle, signIn, signOut } = SvelteKitAuth({
+  trustHost: true,
+  secret: authjsSecret,
+  providers: [Keycloak(kcConfig)],
+});
+```
+
+6. Next, let's review the hooks (`src/hooks.server.ts`) file. This imports the `handle` method by the `src/auth.ts` and re-exports it. This allows the hooks to act as a middleware and include the authentication status of the user in each request.
+
+```tsx
+export { handle } from "./auth.js";
+```
+
+7. To retrieve the authentication status of the user in server-side rendering on the index route, it's using the `+layout.server.ts` and `+page.svelte` files. The `+layout.server.ts` retrieves the authentication status via the `getSession` function from [SvelteKit Locals](https://kit.svelte.dev/docs/types#app-locals) set by `src/hook.server.ts`.
+
+```tsx
+import type { LayoutServerLoad } from "./$types.js";
+
+export const load: LayoutServerLoad = async (event) => {
+  const session = await event.locals.getSession();
+  return {
+    session,
+  };
+};
+```
```

Comment:
@rishi-raj-jain let's add the next few steps following the data through the application: 

Once layout sets that value, +page uses it 

```
<User
        data={{
          user: data?.session?.user,
          status: Boolean(data?.session),
        }}
      />
```

The user component can then determine whether to show Sign In or Out based on the value of the user object.

Reply:
got it adding